### PR TITLE
Granular performance tracking

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

WIP: NO INTENDED TO MERGE YET. 

**Instructions:** 
- To track performance in telemetry, enable the permission `track_performance`.

**To Dos**
- Add a couple of unit tests for the performance service 
- Verify if we want to track all these telemetry records or just the ones with more value. 

**Telemetry:**
| Telemetry | Description |
|----|----|
| perf:<_page_>:apdex:<_score_> |  Count of page loads. Current T is 3, with a tolerance of 4. When load time is <= T then `score` is `satisfied`. When load time is > T but <= 4xT then `score` is `tolerating`. When load time is > 4xT then `score` is `frustrated`.  |
| perf:select_contact:<contact_type>:load_everything | Time taken to fetch from the local device storage all the related data of a selected contact. |
| perf:select_contact:<contact_type>:load_everything:load_contact_data | Time taken to get a contact's data  |
| perf:select_contact:<contact_type>:load_everything:load_children | Time taken to get a contact's descendants places |
| perf:select_contact:<contact_type>:load_everything:load_reports | Time taken to get a contact's associated reports |
| perf:select_contact:<contact_type>:load_everything:load_targets | Time taken to get a contact's associated targets |
| perf:select_contact:<contact_type>:load_everything:load_tasks | Time taken to get a contact's associated tasks |
| perf:select_contact:<contact_type>:load_everything:load_contact_summary | Time taken to get its contact summary |
| perf:select_contact:<contact_type>:display_contact_forms | Time taken to get the contacts forms that are displayed in the FAB |
| perf:select_contact:<contact_type>:display_app_forms| Time taken to get the app forms that are displayed in the FAB |
| perf:contact-list:initial_load | Time taken to get all the data and have it ready to display |
| perf:contact-list:initial_load:fetch_user_place_summary | Time taken to get the user's place summary data |
| perf:contact-list:query:format_contacts | Time taken to format the list of contacts data |
| perf:contact-list:initial_load:fetch_children | Time taken to get the contacts' descendants places |
| perf:contact-list:query:execute_search | Time taken to execute the search |
| perf:contact-list:display_contact_forms | Time taken to prepare the data after getting it from the device's local storage and before rendering |
| perf:forms_by_type:fetch | Time taken to fetch forms from the device's local storage |
| perf:forms_by_type:filter | Time taken to filter forms after fetching them from the device's local storage |

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:track-performance/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:track-performance/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:track-performance/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
@Benmuiruri, these are some places where I put the track. Do you think there is a missing relevant place where we should put it based on yesterday conversation with the team?

File: webapp/src/ts/modules/contacts/contacts.component.ts [hunk 1]
```diff
@@ -230,8 +240,9 @@ export class ContactsComponent implements OnInit, AfterViewInit, OnDestroy {
     return this.authService.has('can_view_last_visited_date');
   }
 
-  private formatContacts(contacts) {
-    return contacts.map(updatedContact => {
+  private formatContacts(contacts) { //NOSONAR
+    const trackPerformance = this.performanceService.track('contact-list:query:format_contacts');
+    const formattedContacts = contacts.map(updatedContact => { //NOSONAR
```
Comment (latin-panda) [inline webapp/src/ts/modules/contacts/contacts.component.ts:245]: Not the scope of this work, I created[ a ticket for that ](https://github.com/medic/cht-core/issues/8803)

**Comment by latin-panda** [comment]
@Benmuiruri @michaelkohn 
Example of the telemetry that this code produces at this point. The telemetry entry prefix is `perf:`
Note that this data is not real testing. It is just my local environment, with little docs and decent internet. Once the CI is green, I can push this branch to our test instance and iterate from there.  

<details>
<summary> Sample of telemetry this produces</summary>

```
{
  "_id": "telemetry-2024-1-11-retta-b9ce8fc4-a6c5-47e9-977b-0962f1e36617",
  "_rev": "1-ccbd84ac812d883c221ca13f758a273c",
  "type": "telemetry",
  "metrics": {
    "boot_time": {
      "sum": 3561.5000000037253,
      "min": 1011.1000000014901,
      "max": 2550.400000002235,
      "count": 2,
      "sumsqr": 7526863.370014414
    },
    "boot_time:1:to_first_code_execution": {
      "sum": 1119.6000000014901,
      "min": 244.5,
      "max": 875.1000000014901,
      "count": 2,
      "sumsqr": 825580.260002608
    },
    "boot_time:2:to_bootstrap": {
      "sum": 2351.89999999851,
      "min": 89.09999999776483,
      "max": 2262.800000000745,
      "count": 2,
      "sumsqr": 5128202.650002974
    },
    "boot_time:2_1:to_replication": {
      "sum": 2226.89999999851,
      "min": 2226.89999999851,
      "max": 2226.89999999851,
      "count": 1,
      "sumsqr": 4959083.609993364
    },
    "boot_time:2_3:to_purge_meta": {
      "sum": 26.799999997019768,
      "min": 26.799999997019768,
      "max": 26.799999997019768,
      "count": 1,
      "sumsqr": 718.2399998402595
    },
    "boot_time:3:to_angular_bootstrap": {
      "sum": 90.00000000372529,
      "min": 43.100000001490116,
      "max": 46.900000002235174,
      "count": 2,
      "sumsqr": 4057.220000338107
    },
    "boot_time:purging_meta:false": {
      "sum": 1,
      "min": 1,
      "max": 1,
      "count": 1,
      "sumsqr": 1
    },
    "boot_time:purging_meta:true": {
      "sum": 1,
      "min": 1,
      "max": 1,
      "count": 1,
      "sumsqr": 1
    },
    "perf:contact-list:display_contact_forms": {
      "sum": 6.300000004470348,
      "min": 1.4000000022351742,
      "max": 1.8999999985098839,
      "count": 4,
      "sumsqr": 10.090000011622905
    },
    "perf:contact-list:initial_load": {
      "sum": 1044.6000000052154,
      "min": 84,
      "max": 457.9000000022352,
      "count": 4,
      "sumsqr": 398893.46000404295
    },
    "perf:contact-list:initial_load:fetch_children": {
      "sum": 0.8000000007450581,
      "min": 0.19999999925494194,
      "max": 0.20000000298023224,
      "count": 4,
      "sumsqr": 0.16000000029802325
    },
    "perf:contact-list:initial_load:fetch_user_place_summary": {
      "sum": 247.10000000149012,
      "min": 37.5,
      "max": 89,
      "count": 4,
      "sumsqr": 17585.11000011355
    },
    "perf:contact-list:query": {
      "sum": 978.3999999985099,
      "min": 21.599999997764826,
      "max": 374.1000000014901,
      "count": 6,
      "sumsqr": 278945.2199999775
    },
    "perf:contact-list:query:format_contacts": {
      "sum": 0.29999999701976776,
      "min": 0,
      "max": 0.10000000149011612,
      "count": 6,
      "sumsqr": 0.029999999403953564
    },
    "perf:forms_by_type:fetch": {
      "sum": 587.6999999992549,
      "min": 266.80000000074506,
      "max": 320.8999999985099,
      "count": 2,
      "sumsqr": 174159.0499994412
    },
    "perf:forms_by_type:filter": {
      "sum": 1872.5000000037253,
      "min": 15.199999999254942,
      "max": 219.80000000074506,
      "count": 38,
      "sumsqr": 163991.41000013362
    },
    "perf:select_contact:display_app_forms": {
      "sum": 20.50000000372529,
      "min": 1.4000000022351742,
      "max": 1.8999999985098839,
      "count": 13,
      "sumsqr": 32.590000007897615
    },
    "perf:select_contact:display_contact_forms": {
      "sum": 21.099999994039536,
      "min": 1.300000000745058,
      "max": 2.399999998509884,
      "count": 13,
      "sumsqr": 35.289999980330464
    },
    "perf:select_contact:load_everything": {
      "sum": 177809.39999998733,
      "min": 690.0999999977648,
      "max": 32160.89999999851,
      "count": 13,
      "sumsqr": 3545718945.43959
    },
    "perf:select_contact:load_everything:contact_data": {
      "sum": 806.3999999985099,
      "min": 28.100000001490116,
      "max": 267.3999999985099,
      "count": 13,
      "sumsqr": 97386.2399992174
    },
    "perf:select_contact:load_everything:load_children": {
      "sum": 681.0000000037253,
      "min": 19.099999997764826,
      "max": 270.1000000014901,
      "count": 13,
      "sumsqr": 96203.68000108944
    },
    "perf:select_contact:load_everything:load_contact_summary": {
      "sum": 98.69999998807907,
      "min": 4.899999998509884,
      "max": 19.799999997019768,
      "count": 13,
      "sumsqr": 1053.7499997213483
    },
    "perf:select_contact:load_everything:load_reports": {
      "sum": 354.29999999701977,
      "min": 10.100000001490116,
      "max": 123.89999999850988,
      "count": 13,
      "sumsqr": 21000.069999587835
    },
    "perf:select_contact:load_everything:load_targets": {
      "sum": 77.8999999947846,
      "min": 3,
      "max": 10.399999998509884,
      "count": 13,
      "sumsqr": 531.929999934584
    },
    "perf:select_contact:load_everything:load_tasks": {
      "sum": 199.39999999850988,
      "min": 1.5,
      "max": 98,
      "count": 13,
      "sumsqr": 12058.540000003428
    },
    "perf:select_contact:set_child_types": {
      "sum": 12.399999994784594,
      "min": 0.5,
      "max": 1.800000000745058,
      "count": 13,
      "sumsqr": 14.959999991357327
    },
    "replication:medic:from:docs": {
      "sum": 4,
      "min": 1,
      "max": 1,
      "count": 4,
      "sumsqr": 4
    },
    "replication:medic:from:ms-since-last-replicated-date": {
      "sum": 27414,
      "min": 5769,
      "max": 12316,
      "count": 3,
      "sumsqr": 271995458
    },
    "replication:medic:from:success": {
      "sum": 673,
      "min": 127,
      "max": 279,
      "count": 4,
      "sumsqr": 129639
    },
    "replication:medic:to:docs": {
      "sum": 0,
      "min": 0,
      "max": 0,
      "count": 4,
      "sumsqr": 0
    },
    "replication:medic:to:ms-since-last-replicated-date": {
      "sum": 27105,
      "min": 5566,
      "max": 12265,
      "count": 3,
      "sumsqr": 267417657
    },
    "replication:medic:to:success": {
      "sum": 353,
      "min": 47,
      "max": 202,
      "count": 4,
      "sumsqr": 48429
    },
    "replication:meta:sync:docs": {
      "sum": 17,
      "min": 0,
      "max": 9,
      "count": 3,
      "sumsqr": 145
    },
    "replication:meta:sync:success": {
      "sum": 722,
      "min": 161,
      "max": 348,
      "count": 3,
      "sumsqr": 192394
    },
    "replication:user-initiated": {
      "sum": 2,
      "min": 1,
      "max": 1,
      "count": 2,
      "sumsqr": 2
    },
    "rules-engine:initialize": {
      "sum": 250,
      "min": 121,
      "max": 129,
      "count": 2,
      "sumsqr": 31282
    },
    "rules-engine:tasks:dirty-contacts": {
      "sum": 0,
      "min": 0,
      "max": 0,
      "count": 4,
      "sumsqr": 0
    },
    "rules-engine:tasks:some-contacts": {
      "sum": 171,
      "min": 23,
      "max": 94,
      "count": 4,
      "sumsqr": 10831
    },
    "rules-engine:tasks:some-contacts:queued": {
      "sum": 0,
      "min": 0,
      "max": 0,
      "count": 4,
      "sumsqr": 0
    },
    "search:contacts:search": {
      "sum": 154.30000000074506,
      "min": 154.30000000074506,
      "max": 154.30000000074506,
      "count": 1,
      "sumsqr": 23808.490000229925
    },
    "search:contacts:types": {
      "sum": 535.0999999977648,
      "min": 9.800000000745058,
      "max": 275.79999999701977,
      "count": 5,
      "sumsqr": 117909.9299983707
    },
    "search:reports": {
      "sum": 89.19999999925494,
      "min": 6.600000001490116,
      "max": 82.59999999776483,
      "count": 2,
      "sumsqr": 6866.319999650418
    },
    "search:reports:subjectIds": {
      "sum": 323.8999999985099,
      "min": 7.899999998509884,
      "max": 121.19999999925494,
      "count": 13,
      "sumsqr": 19318.229999889878
    },
    "user_settings:language:en": {
      "sum": 2,
      "min": 1,
      "max": 1,
      "count": 2,
      "sumsqr": 2
    }
  },
  "device": {
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "hardwareConcurrency": 8,
    "screen": {
      "width": 1792,
      "height": 1044
    },
    "deviceInfo": {}
  },
  "metadata": {
    "year": 2024,
    "month": 1,
    "day": 11,
    "user": "retta",
    "deviceId": "b9ce8fc4-a6c5-47e9-977b-0962f1e36617",
    "versions": {
      "app": "4.5.0-local-development",
      "forms": {
        "approve_mute_household": "23-3e965a0feefc83e8cb3832ce98efdac3",
        "approve_mute_person": "23-ba23c3e3e5d6261e04342ae4a11be3cb",
        "cha_signal_reporting_verification": "23-6862ce325462e5938097b82ba6d27c72",
        "cha_signal_verification": "23-75b8f13e5b0c5ddfb88a8919e809ab39",
        "cha_supervision_calendar": "23-0896b5188198112fd9ca98449e9a38d8",
        "cha_verbal_autopsy": "23-3276ec57e9f61522dafbe84163e7e5d1",
        "chu_supervision": "23-41354974368bd7480d3b537287966c3f",
        "chv_signal_reporting": "23-8a40ba6b535855852ba38ec8452142b0",
        "chv_supervision": "23-20f5d55cbb97525d7a617e214158e678",
        "commodities_order": "23-8319a751ec19d5635a2180fa0fc1e925",
        "commodity_count": "23-ce1aebae3abbf2855beb36ca406d0452",
        "commodity_discrepancy": "23-7b5ef91d5e635c49876fb2323bb0aeb8",
        "commodity_received": "23-4dc537c9c2df0024285093c33360acf7",
        "commodity_return": "23-d826d6ca25a8dd9b68238646e120b191",
        "commodity_returned": "23-eb34cafcaf1f14fbc6ccdba3e448751c",
        "commodity_stockout": "23-68235ab5924b09c6503ac76eaa62933d",
        "commodity_supply": "23-790a6ee22012fa2b4e93d3433fa22f2a",
        "community_event": "23-06ed3188c2e9420fd72e862160955c9b",
        "contact:a_county:create": "23-743a76429afdb75838cb108e4cbd2897",
        "contact:a_county:edit": "23-eb9f0d35665fb221ecb53328b78bf61e",
        "contact:b_sub_county:create": "23-5498add0d4ec7cb6827e27296e0743b6",
        "contact:b_sub_county:edit": "23-ec604b5dd342635b9833ef3f72a6ee8f",
        "contact:c_community_health_unit:create": "23-f1aad24c95e9b70021eb00bde87048da",
        "contact:c_community_health_unit:edit": "23-d9c2558d58be363fdb50a87af380e91e",
        "contact:d_community_health_volunteer_area:create": "23-6d18e3807dcddc37c8ec5295ebf7b4eb",
        "contact:d_community_health_volunteer_area:edit": "23-002c904313b7f97e71de8166fa7eb7b5",
        "contact:e_household:create": "23-050bb13f1299f18838ce425bb24e6d58",
        "contact:e_household:edit": "23-7def261a6a64a4cdec8954d3dccfce5a",
        "contact:f_client:create": "23-74b613460c34d96671893c128072088c",
        "contact:f_client:edit": "23-40e6ed527c93d6dc2f80b8efd222dd7d",
        "contact:person:create": "26-375e656fd0d0c525dfe5e8411a40dd4d",
        "contact:person:edit": "26-ce909f30a2ec23309ab77bfb153cacfc",
        "death_report": "26-87730ec3fdba44df2a9a674b95a91a11",
        "death_review": "23-087f3290f5ab018bb31def716d50b3c0",
        "defaulter_follow_up": "23-af5773da83e484460b2b474020b5257d",
        "family_planning": "23-d4226eb4d39d49519746711e63c024c0",
        "household_member_registration_reminder": "23-0eb14edd789d7a8a7386d278dc265b3a",
        "immunization_service": "23-42db08690482f560843a19f3f0cd4291",
        "mute_household": "23-1c4887e28c007b54d9ff66b29100e642",
        "mute_person": "23-8a2135d9124977502d447d683a6eede3",
        "over_five_assessment": "23-e04930acbed4b2d0fae258b60210d06c",
        "patient_details_reminder": "23-08999bcb16cb9166001f8746dcef4504",
        "postnatal_care_service": "23-a4994d41819d2beed332626c2d6402b4",
        "postnatal_care_service_newborn": "23-b0fc176f7a202401677b82f4f89ad979",
        "pregnancy_home_visit": "26-67fec73a65c4eb48f7dbd0d509edb576",
        "referral_follow_up": "23-9c76b9184ed551b865691d12890ae5c0",
        "sgbv": "23-ab9a247104f290f59134a1fc2f251c4e",
        "socio_economic_survey": "23-54ec72d83cdc976877cd073f5e11be8e",
        "socio_economic_survey_person": "23-157072ee934a8b7572135510c7a8449b",
        "treatment_follow_up": "23-834e8095a1e6e7f486950b11f731a447",
        "u5_assessment": "23-d45f77bb4830d907af1acc6624205d66",
        "unmute_household": "23-cba7d6f8619343600b65fa14512594a7",
        "unmute_person": "23-7822827165c538c72ea5148815b5b4ba",
        "wash": "23-aa01cb70c85b1e8276559b450a2257d4"
      },
      "settings": "3-b4fed6d407ce677c7de0cfb8cca8af96"
    }
  },
  "dbInfo": {
    "doc_count": 83,
    "update_seq": 83,
    "idb_attachment_format": "binary",
    "db_name": "medic-user-retta",
    "auto_compaction": true,
    "adapter": "idb"
  }
}

```

</details>

File: webapp/src/ts/effects/contacts.effects.ts [hunk 2]
```diff
@@ -98,6 +104,7 @@ export class ContactsEffects {
   }
 
   private loadContact(id) {
+    const trackPerformance = this.performanceService.track('select_contact:load_everything:contact_data');
```
Comment (Benmuiruri) [inline webapp/src/ts/effects/contacts.effects.ts:107]: How about 'select_contact:load_everything:load_contact_data' for consistency ?

File: webapp/src/ts/modules/contacts/contacts.component.ts [hunk 3]
```diff
@@ -315,6 +335,7 @@ export class ContactsComponent implements OnInit, AfterViewInit, OnDestroy {
   }
 
   private query(opts?) {
+    const trackPerformance = this.performanceService.track('contact-list:query');
```
Comment (Benmuiruri) [inline webapp/src/ts/modules/contacts/contacts.component.ts:338]: How about 'contact-list:query: execute_search' 

Also should we move it to the [search](https://github.com/medic/cht-core/blob/10ff96d98d8d3bdd1dfa97a47ae82f0403b4a105/webapp/src/ts/modules/contacts/contacts.component.ts#L431C3-L431C13) method which performs the actual search ?

**Review by Benmuiruri — COMMENTED** [review]
@latin-panda I think you've covered enough for a start, for us to get things going. Later we might consider adding some more telemetry in other contact related activities such as editing a contact. But for now I think we have enough for testing the contact component.

File: webapp/src/ts/effects/contacts.effects.ts (hunk 2 above)
Comment (latin-panda) [reply webapp/src/ts/effects/contacts.effects.ts:107]: Ah makes sense! thanks

File: webapp/src/ts/modules/contacts/contacts.component.ts (hunk 3 above)
Comment (latin-panda) [reply webapp/src/ts/modules/contacts/contacts.component.ts:338]: > How about 'contact-list:query: execute_search'

Done

> Also should we move it to the search method which performs the actual search?

That function has 2 paths one executes the query almost right away, the other does a navigation, which I think ends up trigger the search from the ngoninit, and it already have a track: `contact-list:initial_load`

**Comment by michaelkohn** [comment]
These are great! I like how they are kind of "nested" so it's clearer which ones are a subset of others. Based on the descriptions, I tried to figure out what that translates to in the UI. I've documented that in this deck and have some questions. I think ultimately the two that we will really care about for Apdex will be `select_contact:load_everything` and `contact-list:initial_load`. 

https://docs.google.com/presentation/d/1CN3QTu-uNRDXJKsi5HxiYST19zmScV31k6gCE3Q8mp4/edit#slide=id.gdb697cfd5c_0_69

**Review by Benmuiruri — APPROVED** [review]
Looks ready to start logging the data.

**Comment by latin-panda** [comment]
@Benmuiruri  I've opened a PR for merging the [Apdex](https://github.com/medic/cht-core/pull/8822).
I'll keep this one open to see where we are adding the granular tracking that we are deploying in the test server.

**Comment by latin-panda** [comment]
Closing in favor of this [other updated version](https://github.com/medic/cht-core/pull/8829)

**Pull request closed without merging**